aten/src/ATen/core/type.cpp

```diff
@@ -141,73 +141,78 @@ ListTypePtr ListType::ofBools() {
-  if (t1->isSubtypeOf(NoneType::get()) && !t2->isSubtypeOf(NoneType::get())) {
-    return OptionalType::create(t2);
-  } else if (t2->isSubtypeOf(NoneType::get()) && !t1->isSubtypeOf(NoneType::get())) {
-    return OptionalType::create(t1);
-  }
+    //cases that t1 == t2, or t1 is a type refinement of t2 and vice versa
+    if (t1->isSubtypeOf(t2)) {
+      return t2;
+    } else if (t2->isSubtypeOf(t1)) {
+      return t1;
+    }
 
-  //types which contain other types
-  if (t1->cast<ListType>() && t2->cast<ListType>()) {
-    // because we have runtime specializations of lists, e.g. int[] = std::vector<int64_t>
-    // int?[] = std::vector<IValue>  we don't allow type coercion,
-    // since t1 & t2 may have different runtime representations.
-
-    // allow Lists of different tensor types
-    auto unshaped_t1 = unshapedType(t1);
-    auto unshaped_t2 = unshapedType(t2);
-    return tryEitherIsTheSuperType(unshaped_t1, unshaped_t2);
-  } else if(t1->cast<TupleType>() && t2->cast<TupleType>()) {
-    auto tuple1 = t1->cast<TupleType>();
-    auto tuple2 = t2->cast<TupleType>();
-    if (tuple1->elements().size() != tuple2->elements().size()) {
-      return c10::nullopt;
+    // if t1 is None and t2 is a concrete type, return Optional[t2] and vice versa
+    if (t1->isSubtypeOf(NoneType::get()) && !t2->isSubtypeOf(NoneType::get())) {
+      return OptionalType::create(t2);
+    } else if (t2->isSubtypeOf(NoneType::get()) && !t1->isSubtypeOf(NoneType::get())) {
+      return OptionalType::create(t1);
     }
-    std::vector<TypePtr> elements;
-    for (size_t i = 0; i < tuple1->elements().size(); i++) {
-      if (auto elem = unifyTypes(tuple1->elements().at(i), tuple2->elements().at(i))) {
-        elements.push_back(*elem);
-      } else {
+
+    if (t1->cast<TupleType>() && t2->cast<TupleType>()) {
+      auto tuple1 = t1->cast<TupleType>();
+      auto tuple2 = t2->cast<TupleType>();
+      if (tuple1->elements().size() != tuple2->elements().size()) {
         return c10::nullopt;
       }
+      std::vector<TypePtr> elements;
+      for (size_t i = 0; i < tuple1->elements().size(); i++) {
+        if (auto elem = unifyTypes(tuple1->elements().at(i), tuple2->elements().at(i))) {
+          elements.push_back(*elem);
+        } else {
+          return c10::nullopt;
+        }
+      }
+      return static_cast<TypePtr>(TupleType::create(elements));
     }
-    return static_cast<TypePtr>(TupleType::create(elements));
-  } else if (t1->cast<DictType>() && t2->cast<DictType>()) {
+  }
+
+  // NB: we do not return NumberType because there is not currently enough
+  // operator support for it
+
+  // because we have runtime specializations of lists, e.g. int[] = std::vector<int64_t>
+  // int?[] = std::vector<IValue>  we don't allow type coercion,
+  // since t1 & t2 may have different runtime representations. if we did, we would
+  // we would try to unify the List element type.
+  // as is, List is covered by EitherIsTheSuperType call above
```

Comment:
This call is gone now, right?

Reply:
yep, good catch , just updated

Reply:
> @eellison your edit might mean you agree with this already, but I think Zach's rule is the right one bc `Optional` is the only case where the nesting level is unequal between the two types.

Yep, agreed. Just pushed update, I think it is correct now. Could you take a look now because i'm not sure of my comments?